hooks/tk-multi-publish2/basic/nuke/publish_files.py

```diff
@@ -0,0 +1,182 @@
+        """
+        Construction
+        """
+        # call base init
+        super(NukePublishFilesCustomPlugin, self).__init__(parent, **kwargs)
+
+    def _sync_frame_range(self):
+        """
+        Checks whether frame range is in sync with shotgun.
+
+        :return: True if yes false otherwise
+        """
+        # get the current engine
+        engine = sgtk.platform.current_engine()
+        # context from engine
+        context = engine.context
+        entity = context.entity
+
+        sg_entity_type = context.entity["type"]
+        sg_filters = [["id", "is", entity["id"]]]
+        fields = ["cut_in", "cut_out"]
+
+        # get the field information from shotgun based on Shot
+        data = engine.shotgun.find_one(sg_entity_type, filters=sg_filters, fields=fields)
+        # compare if the frame range set at root level is same as the shogun cut_in, cut_out
+        root = nuke.Root()
+        if root.firstFrame() != data["cut_in"] or root.lastFrame() != data["cut_out"]:
+            self.logger.error("Frame range not synced with Shotgun.")
+            return False
+        return True
+
+    def _non_sgtk_writes(self):
+        """
+        Checks for non SGTK write nodes present in the scene.
+
+        :return: True if yes false otherwise
+        """
+        write_nodes = ""
+        # get all write nodes
+        write = nuke.allNodes('Write')
+        if write:
+            for item in range(len(write)):
+                write_nodes += "\n" + write[item]['name'].value()
+            self.logger.error("Non SGTK write nodes detected.",
+                              extra={
+                                  "action_show_more_info": {
+                                      "label": "Show Info",
+                                      "tooltip": "Show non sgtk write node(s)",
+                                      "text": "Non SGTK write nodes:\n{}".format(write_nodes)
+                                  }
+                              }
+                              )
+            return False
+        return True
+
+    @staticmethod
+    def _check_for_knob(node, knob):
+        try:
+            node[knob]
+        except NameError:
+            return False
+        else:
+            return True
+
+    def _file_paths(self):
+        """
+        Checks if the files loaded are from valid paths i.e
+        /dd/shows/<show_name> or /dd/library
+
+        :return: True if paths are valid false otherwise
+        """
+        show = os.environ['DD_SHOW']
+        valid_paths = ['/dd/shows/' + show, '/dd/library']
+        paths = ""
+
+        # For all the nodes present in the script, check for 'file' parameter.
+        # If its populated, get the file path and check validity.
+        for index, fileNode in enumerate(nuke.allNodes(), 0):
+            # if the node has 'file' parameter, check if the file path is valid
+            if self._check_for_knob(fileNode, 'file'):
```

Comment:
Actually, item.properties['path'] is the output path. You will need to traverse up the graph from the output node and get all input paths, and check them as well. I believe there is a way to traverse the graph via the nuke api.

Reply:
Maybe I'm not very clear here. The file paths being checked here are the ones inside the nuke script like the ones used by read or camera node. I wasn't sure what all nodes have 'file' parameter so was moving through all the nodes but would be better to just keep it to read, readgeo and camera nodes.

Reply:
This process operates on a WriteNode, that is what is stored in item.properties['node']. From that WriteNode, you should be traversing up the node graph to get any/all inputs that have a file parameter, and then checking those. At present, you are reading all read nodes in the scene regardless of whether or not they correlate to the WriteNode being exported. This will result in false positives as well as redundant warnings if there is more than one WriteNode.